### PR TITLE
HPCC-27887 Implement new keyed compression format

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jhtree/ctfile.hpp [hunk 1]
```diff
@@ -92,11 +98,12 @@ struct __declspec(novtable) jhtree_decl KeyHdr
...
     short nodeKeyLength; /* key length in intermediate level nodes e8x */
     unsigned short version; /* build version - to be updated if key format changes    eax*/
-    short unused[2]; /* unused ecx */
+    unsigned keyFlags; /* extended version of ktype ecx */
     __int64 blobHead; /* fpos of first blob node f0x */
     __int64 metadataHead; /* fpos of first metadata node f8x */
     __int64 bloomHead; /* fpos of bloom table data, if present 100x */
-    __uint64 partitionFieldMask; /* Bitmap indicating partition keyed fields */
+    __uint64 partitionFieldMask; /* Bitmap indicating partition keyed fields 108x */
+    unsigned getFlags() { return keyFlags; }
```
Comment (richardkchapman) [inline system/jhtree/ctfile.hpp:106]: Won't this fail on an old index? Do you need to fall back to returning oldktype if keyFlags is (whatever value we can be sure is in old index files).

File: system/jhtree/ctfile.hpp [hunk 2]
```diff
@@ -161,14 +185,15 @@ class jhtree_decl CKeyHdr : public CWritableKeyNode
...
-    inline char getKeyType() { return hdr.ktype; }
+    inline unsigned getKeyType() { return hdr.keyFlags; }
     inline offset_t getRootFPos() { return hdr.root; }
     inline unsigned short getMaxNodeBytes() { return hdr.maxkbl; }
     inline KeyHdr *getHdrStruct() { return &hdr; }
     inline static size32_t getSize() { return sizeof(KeyHdr); }
     inline unsigned getNodeSize() { return hdr.nodeSize; }
     inline bool hasSpecialFileposition() const { return true; }
-    inline bool isRowCompressed() const { return (hdr.ktype & (HTREE_QUICK_COMPRESSED_KEY|HTREE_VARSIZE)) == HTREE_QUICK_COMPRESSED_KEY; }
+    inline bool isRowCompressed() const { return (hdr.keyFlags & (HTREE_QUICK_COMPRESSED_KEY|HTREE_VARSIZE)) == HTREE_QUICK_COMPRESSED_KEY; }
```
Comment (richardkchapman) [inline system/jhtree/ctfile.hpp:195]: Similar compatibility issues...

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Some initial concerns from first (very quick) scan

File: system/jhtree/keybuild.cpp [hunk 3]
```diff
@@ -168,11 +170,24 @@ class CKeyBuilder : public CInterfaceOf<IKeyBuilder>
                     bloomInfo++;
                 }
             }
+
+            //Create a representation of the null row - used for the new compression algorithm
```
Comment (richardkchapman) [inline system/jhtree/keybuild.cpp:174]: Should this be conditional?

File: system/jhtree/keybuild.cpp [hunk 4]
```diff
@@ -119,6 +114,7 @@ class CKeyBuilder : public CInterfaceOf<IKeyBuilder>
     CRC32EndHT crcEndPosTable;
     CRC32 headCRC;
     bool doCrc = false;
+    KeyBuildContext ctx;
```
Comment (richardkchapman) [inline system/jhtree/keybuild.cpp:117]: Is this used?

File: system/jhtree/jhinplace.hpp [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+{
+public:
+    InplaceIndexCompressor(size32_t keyedSize, IHThorIndexWriteArg * helper);
+    virtual ~InplaceIndexCompressor();
+
+    virtual const char *queryName() const override { return "inplace"; }
+    virtual CWriteNode *createNode(offset_t _fpos, CKeyHdr *_keyHdr, bool isLeafNode) const override
+    {
+        if (isLeafNode)
+            return new CInplaceLeafWriteNode(_fpos, _keyHdr, const_cast<KeyBuildContext &>(ctx));
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.hpp:243]: cast seems unnecessary/confusing

File: system/jhtree/jhinplace.cpp [hunk 6]
```diff
@@ -0,0 +1,2681 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:3]: Wrong date

File: system/jhtree/jhinplace.cpp [hunk 7]
```diff
@@ -0,0 +1,2681 @@
...
+    //SqMatchX                    // Matches a string at offset (cur-1-count) - delta8/16.  Only likely to help with repeated trailing key segments
+};
+
+//inline unsigned reverseBytes(unsigned value) { return __builtin_bswap32(value); }
+
+//MORE: What other special values are worth encoding in real life indexes?
+static constexpr byte specialSingleChars[31] = {
+    0, 1, 2, 3, 4, 5, 6, 7,                 // 1-8      Useful for small numbers
+    ' ', '0', 'F', 'M', 'U',                // 9-13     Single padding and gender
+                                            // 14-31    Currently unused.
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:72]: Other digits would seem likely to occur?

File: system/jhtree/jhinplace.cpp [hunk 8]
```diff
@@ -0,0 +1,2681 @@
...
+    {
+        byte next = data.bytes()[i];
+        dataHex.appendhex(next, true);
+        if (isprint(next))
+            clean.append(next);
+        else
+            clean.append('.');
+    }
+
+    printf("%*s(%s[%s] %u:%u[%u] [%s]", indent, "", clean.str(), dataHex.str(), data.length(), squashedData.length(), getSize(), squashedText.str());
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:929]: DBGLOG would be more normal...

File: system/jhtree/jhinplace.cpp [hunk 9]
```diff
@@ -0,0 +1,2681 @@
...
+        printf("%*s}", indent, "");
+    }
+    printf(")\n");
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+
+void PartialMatchBuilder::add(size32_t len, const void * data)
+{
+    //It is legal to add rows longer than keyLen, but cannot strip trailing nulls
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:944]: Why would you want to?

File: system/jhtree/jhinplace.cpp [hunk 10]
```diff
@@ -0,0 +1,2681 @@
...
+
+// Only used for counts, so performance is less critical - implement fully once everything else is implemented
+int CJHInplaceTreeNode::locateGT(const char * search, unsigned minIndex) const
+{
+    unsigned int a = minIndex;
+    int b = getNumKeys();
+    // Locate first record greater than src
+    while ((int)a<b)
+    {
+        //MORE: Note sure why the index is subtly different to the GTE version
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:1641]: typo: Note -> not

File: system/jhtree/jhinplace.cpp [hunk 11]
```diff
@@ -0,0 +1,2681 @@
...
+
+        firstSequence = readPacked64(data);
+        positionData = data;
+
+        if (bytesPerPosition != 0)
+            data += (bytesPerPosition * (numKeys -1));
+
+        if (isLeaf())
+        {
+            //Extra position serialized for leafs..
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:1689]: leaves? leaf nodes?

File: system/jhtree/jhinplace.cpp [hunk 12]
```diff
@@ -0,0 +1,2681 @@
...
+            const RtlFieldInfo *field = meta.queryField(idx);
+            offset = field->type->buildNull(rowBuilder, offset, field);
+        }
+        ctx.nullRow = nullRow;
+    }
+}
+
+InplaceIndexCompressor::~InplaceIndexCompressor()
+{
+    delete [] ctx.nullRow;
```
Comment (richardkchapman) [inline system/jhtree/jhinplace.cpp:2162]: Shouldn't ctc destructor be responsible for this?

**Review by richardkchapman — COMMENTED** [review]
Some initial comments from a quick scan

File: system/jhtree/jhinplace.hpp (hunk 5 above)
Comment (ghalliday) [reply system/jhtree/jhinplace.hpp:243]: changed the member so it is mutable - much cleaner

File: system/jhtree/keybuild.cpp (hunk 4 above)
Comment (ghalliday) [reply system/jhtree/keybuild.cpp:117]: no, historic location prior to refactoring

File: system/jhtree/jhinplace.cpp (hunk 7 above)
Comment (ghalliday) [reply system/jhtree/jhinplace.cpp:72]: This is something that would be good to tune on some real-life indexes.

File: system/jhtree/jhinplace.cpp (hunk 9 above)
Comment (ghalliday) [reply system/jhtree/jhinplace.cpp:944]: Previous iteration of this code where it was also being used for compressing the payload - will remove the comment.

File: system/jhtree/jhinplace.cpp (hunk 12 above)
Comment (ghalliday) [reply system/jhtree/jhinplace.cpp:2162]: moved

**Review by richardkchapman — APPROVED** [review]
Changes in second commit look good. Worth squashing and merging I think

**Review by richardkchapman — APPROVED** [review]

**Comment by richardkchapman** [comment]
Ditto third commit

**Pull request merged**